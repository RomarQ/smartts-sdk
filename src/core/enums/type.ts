export enum TypeAtom {
    record = 'record',
    variant = 'variant',
    chain_id = 'chain_id',
    unit = 'unit',
    nat = 'nat',
    int = 'int',
    mutez = 'mutez',
    bool = 'bool',
    bytes = 'bytes',
    string = 'string',
    address = 'address',
    timestamp = 'timestamp',
    bls12_381_fr = 'bls12_381_fr',
    bls12_381_g1 = 'bls12_381_g1',
    bls12_381_g2 = 'bls12_381_g2',
    key = 'key',
    key_hash = 'key_hash',
    list = 'list',
    set = 'set',
    option = 'option',
    map = 'map',
    big_map = 'bigmap',
    signature = 'signature',
    operation = 'operation',
    never = 'never',
    tuple = 'tuple',
    lambda = 'lambda',
    ticket = 'ticket',
    sapling_state = 'sapling_state',
    contract = 'contract',
    sapling_transaction = 'sapling_transaction',
}

export default TypeAtom;
