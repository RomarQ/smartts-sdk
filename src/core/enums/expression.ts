export enum ExpressionAtom {
    // Math
    add = 'add',
    mul = 'mul',
    sub = 'sub',
    truediv = 'truediv',
    // Comparison
    eq = 'eq',
    neq = 'neq',
    lt = 'lt',
    gt = 'gt',
    le = 'le',
    ge = 'ge',
    //
    items = 'items',
    update_map = 'update_map',
    getItem = 'getItem',
    getItemDefault = 'getItemDefault',
    getItemMessage = 'getItemMessage',
    attr = 'attr',
    params = 'params',
    getLocal = 'getLocal',
    iter = 'iter',
    // Blockchain property
    amount = 'amount',
    balance = 'balance',
    chain_id = 'chain_id',
    level = 'level',
    now = 'now',
    self = 'self',
    self_address = 'self_address',
    sender = 'sender',
    source = 'source',
    total_voting_power = 'total_voting_power',
    voting_power = 'voting_power',
    // Operation
    operations = 'operations',
    transfer = 'transfer',
    set_delegate = 'set_delegate',
    create_contract = 'create_contract',
    // List instructions
    cons = 'cons',
    //
    contract = 'contract',
    // Variant
    openVariant = 'openVariant',
    isVariant = 'isVariant',
    cases_arg = 'cases_arg',
    variant_arg = 'variant_arg',
}

export default ExpressionAtom;
