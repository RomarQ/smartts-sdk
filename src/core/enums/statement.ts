export enum StatementAtom {
    verify = 'verify',
    ifBlock = 'ifBlock',
    elseBlock = 'elseBlock',
    forGroup = 'forGroup',
    whileBlock = 'whileBlock',
    set = 'set',
    defineLocal = 'defineLocal',
    result = 'result',
    failwith = 'failwith',
    match = 'match',
    match_cases = 'match_cases',
    set_type = 'set_type',
}

export default StatementAtom;
