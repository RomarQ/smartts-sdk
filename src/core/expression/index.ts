export * from './literal';
export * from './expression';
export * from './logic';
