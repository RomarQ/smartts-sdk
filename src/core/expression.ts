import type { IExpression } from '../typings/expression';
import type { ILiteral } from '../typings/literal';
import type { IStatement } from '../typings/statement';
import type { IType } from '../typings/type';
import { GetProperty, LambdaArgument } from '../expression/variables';
import { capitalizeBoolean, LineInfo, parenthesis } from '../misc/utils';
import ValueAtom from './enums/literal';
import TypeAtom from './enums/type';

export class Expression<T extends ValueAtom> implements IExpression<T> {
    _isExpression = true as const;
    // Used for type checking
    _type = {} as T;
    private args;

    constructor(public name: string, ...args: (IExpression | LineInfo | IType | string)[]) {
        this.args = args || [];
    }

    static proxyHandler: ProxyHandler<IExpression> = {
        get: function (target, prop: string, receiver) {
            if (prop in target || typeof prop === 'symbol') {
                return Reflect.get(target, prop, receiver);
            }
            return GetProperty(target, prop);
        },
        apply: function (target, thisArg, argumentsList) {
            console.log(target, thisArg, argumentsList);
            return target;
        },
    };

    toString() {
        return `(${[this.name, ...this.args].join(' ')})`;
    }
}

export class LiteralExpression<T extends ValueAtom> implements ILiteral<T> {
    _isExpression = true as const;
    // Used for type checking
    _type = {} as T;

    constructor(private name: T, private values: (number | string | boolean | IExpression)[], private line: LineInfo) {}

    toString() {
        switch (this.name) {
            case ValueAtom.unit:
                return parenthesis(this.name);
            case ValueAtom.nat:
            case ValueAtom.int:
            case ValueAtom.mutez:
            case ValueAtom.bytes:
            case ValueAtom.bool:
            case ValueAtom.string:
            case ValueAtom.chain_id_cst:
            case ValueAtom.timestamp:
            case ValueAtom.bls12_381_fr:
            case ValueAtom.bls12_381_g1:
            case ValueAtom.bls12_381_g2:
            case ValueAtom.key:
            case ValueAtom.key_hash:
            case ValueAtom.signature:
            case ValueAtom.address:
                return `(literal (${this.name} ${this.values.join(' ')}) ${this.line})`;
            case ValueAtom.list:
            case ValueAtom.set:
            case ValueAtom.tuple:
                return `(${this.name} ${this.line} ${this.values.join(' ')})`;
            case ValueAtom.Some:
            case ValueAtom.None:
                return `(${ValueAtom.variant} "${this.name}" ${this.values.join(' ') || `(${ValueAtom.unit})`} ${
                    this.line
                })`;
            default:
                return `(${this.name} ${this.values.join(' ')} ${this.line})`;
        }
    }
}

export class RecordLiteral implements ILiteral<ValueAtom.record> {
    _isExpression = true as const;
    // Used for type checking
    _type = {} as ValueAtom.record;

    constructor(private fields: Record<string, IExpression>, private line: LineInfo) {}

    private buildFields = (fields: Record<string, IExpression>) => {
        return Object.entries(fields).map(([field, expr]) => `(${field} ${expr.toString()})`);
    };

    toString(): string {
        return `(${TypeAtom.record} ${this.line} ${this.buildFields(this.fields).join(' ')})`;
    }
}

export class MapLiteral<T extends ValueAtom.map | ValueAtom.big_map> implements ILiteral<T> {
    _isExpression = true as const;
    _type = {} as T;
    type = {} as IType;

    constructor(
        private prim: T,
        private rows: IExpression[][],
        keyType: IType,
        valueType: IType,
        private line: LineInfo,
    ) {}

    private buildEntry = ([key, value]: IExpression[]) => {
        return `(${key.toString()} ${value.toString()})`;
    };

    toString() {
        return `(${this.prim} ${this.line} ${this.rows.map(this.buildEntry).join(' ')})`;
    }
}

export class LambdaLiteral implements ILiteral<ValueAtom.lambda> {
    _isExpression = true as const;
    // Used for type checking
    _type = ValueAtom.lambda as const;
    type = {} as unknown as IType;
    static idCounter = 0;
    private identifier: number;
    private withStorage?: 'read-write' | 'read-only';
    private withOperations = false;

    constructor(
        private statements: IStatement[],
        private inType: IType,
        private argumentName: string,
        private line: LineInfo,
    ) {
        this.identifier = ++LambdaLiteral.idCounter;
    }

    public get id(): Readonly<number> {
        return this.identifier;
    }

    public setInputType(type: IType) {
        this.inType = type;
        return this;
    }

    public code(buildStatements: (arg: IExpression) => IStatement[]) {
        const param = LambdaArgument(this.argumentName, this.inType, this.id, this.line);
        this.statements = buildStatements(param);
        return this;
    }

    toString() {
        return `(${ValueAtom.lambda} ${this.id} ${this.withStorage || 'None'} ${capitalizeBoolean(
            this.withOperations,
        )} "${this.argumentName}" ${this.line} (${this.statements.join(' ')}))`;
    }
}
