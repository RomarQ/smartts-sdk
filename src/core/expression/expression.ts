import { proxy } from '../../misc/proxy';
import { LineInfo } from '../../misc/utils';
import { IExpression } from '../../typings/expression';
import { ILiteral } from '../../typings/literal';
import { IType } from '../../typings/type';

export class Expression implements IExpression {
    _isExpression = true as const;
    args;

    constructor(public name: string, ...args: (IExpression | LineInfo | IType | string | ILiteral<unknown>)[]) {
        this.args = args || [];
    }

    static proxyHandler: ProxyHandler<IExpression> = {
        get: function (target, prop: string, receiver) {
            if (prop in target || typeof prop === 'symbol') {
                return Reflect.get(target, prop, receiver);
            }
            return Expression.getAttr(target, prop);
        },
        apply: function (target, thisArg, argumentsList) {
            console.log(target, thisArg, argumentsList);
            return target;
        },
    };

    static getAttr = (from: IExpression, attr: string, line = new LineInfo()) =>
        proxy(new Expression('attr', from, `"${attr}"`, line), Expression.proxyHandler);

    toString() {
        return `(${[this.name, ...this.args].join(' ')})`;
    }
}
