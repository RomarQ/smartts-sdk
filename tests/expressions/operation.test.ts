import { Address, GetContract, GetOperations, Mutez, Transfer, Unit, AppendToList, SetDelegate, Key_hash, Some, None, CreateContract } from '../../src/expression';
import { Contract, EntryPoint } from '../../src/core';
import { verifyContractCompilationOutput } from '../util';
import { SetValue } from '../../src/statement';
import { TKey_hash, TNat } from '../../src/type';

describe('Transaction operation', () => {
    it('Build operation (Desugared version)', () => {
        const contract = new Contract()
            .setStorage(Unit())
            .addEntrypoint(
                new EntryPoint('ep1').code((arg) => [
                    SetValue(
                        GetOperations(),
                        AppendToList(
                            GetOperations(),
                            Transfer(GetContract(Address('tz1gTnKMA65qaKVpp6x4cgMLU2UyKF2zjHYN')), Mutez(100), arg),
                        ),
                    ),
                ]),
            )
            .toString();

        expect(contract).toMatchSnapshot();
        verifyContractCompilationOutput(contract);
    });
    it('Build operation (Sugared version)', () => {
        const contract = new Contract()
            .setStorage(Unit())
            .addEntrypoint(
                new EntryPoint('ep1').code(() => [
                    Transfer(GetContract(Address('tz1gTnKMA65qaKVpp6x4cgMLU2UyKF2zjHYN')), Mutez(100)).send(),
                ]),
            )
            .toString();

        expect(contract).toMatchSnapshot();
        verifyContractCompilationOutput(contract);
    });
    it('Build operation to an originated contract', () => {
        const contract = new Contract()
            .setStorage(Unit())
            .addEntrypoint(
                new EntryPoint('ep1')
                    .inputType(TNat())
                    .code((arg) => [
                        Transfer(
                            GetContract(Address('KT1R9M3MDffw7qSVSnbJs46aMC9YzzZz3aGT'), 'ep1', TNat()),
                            Mutez(100),
                            arg,
                        ).send(),
                    ]),
            )
            .toString();

        expect(contract).toMatchSnapshot();
        verifyContractCompilationOutput(contract);
    });
});

describe('Delegation operation', () => {
    it('Build operation (Desugared version)', () => {
        const contract = new Contract()
            .setStorage(Unit())
            .addEntrypoint(
                new EntryPoint('ep1').code((arg) => [
                    SetValue(
                        GetOperations(),
                        AppendToList(
                            GetOperations(),
                            SetDelegate(Some(Key_hash("tz1gTnKMA65qaKVpp6x4cgMLU2UyKF2zjHYN"))),
                        ),
                    ),
                ]),
            )
            .toString();

        expect(contract).toMatchSnapshot();
        verifyContractCompilationOutput(contract);
    });
    it('Build operation to set delegator (Sugared version)', () => {
        const contract = new Contract()
            .setStorage(Unit())
            .addEntrypoint(
                new EntryPoint('ep1').inputType(TKey_hash()).code((arg) => [
                    SetDelegate(Some(arg)).send(),
                ]),
            )
            .toString();

        expect(contract).toMatchSnapshot();
        verifyContractCompilationOutput(contract);
    });
    it('Build operation to clear the delegation (Sugared version)', () => {
        const contract = new Contract()
            .setStorage(Unit())
            .addEntrypoint(
                new EntryPoint('ep1').code(() => [SetDelegate(None()).send()]),
            )
            .toString();

        expect(contract).toMatchSnapshot();
        verifyContractCompilationOutput(contract);
    });
});

describe('Origination operation', () => {
    it('Build operation (Desugared version)', () => {
        const contract = new Contract()
            .setStorage(Unit())
            .addEntrypoint(
                new EntryPoint('ep1').code((arg) => [
                    SetValue(
                        GetOperations(),
                        AppendToList(
                            GetOperations(),
                            CreateContract(new Contract(), Unit()).getOperation(),
                        ),
                    ),
                ]),
            )
            .toString();

        expect(contract).toMatchSnapshot();
        verifyContractCompilationOutput(contract);
    });
    it('Build operation (Sugared version)', () => {
        const contract = new Contract()
            .setStorage(Unit())
            .addEntrypoint(
                new EntryPoint('ep1').code((arg) => [
                    CreateContract(new Contract(), arg, Mutez(100), Some(Key_hash("tz1gTnKMA65qaKVpp6x4cgMLU2UyKF2zjHYN"))).send()
                ]),
            )
            .toString();

        expect(contract).toMatchSnapshot();
        verifyContractCompilationOutput(contract);
    });
});
