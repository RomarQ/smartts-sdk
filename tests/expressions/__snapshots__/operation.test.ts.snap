// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Delegation operation Build operation (Desugared version) 1`] = `
"
        (
            template_id (static_id 0 (\\"operation.test.ts\\" 63))
            storage (unit)
            storage_type (())
            messages ((ep1 True False False False (\\"operation.test.ts\\" 66) ((set (operations (\\"operation.test.ts\\" 66)) (cons (set_delegate (variant \\"Some\\" (literal (key_hash tz1gTnKMA65qaKVpp6x4cgMLU2UyKF2zjHYN) (\\"operation.test.ts\\" 66)) (\\"operation.test.ts\\" 66)) (\\"operation.test.ts\\" 66)) (operations (\\"operation.test.ts\\" 66)) (\\"operation.test.ts\\" 66)) (\\"operation.test.ts\\" 66)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"operation.test.ts\\" 63))
        )
        "
`;

exports[`Delegation operation Build operation (Desugared version) 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\",
                \\"annots\\": [
                    \\"%ep1\\"
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PUSH\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"option\\",
                            \\"args\\": [
                                {
                                    \\"prim\\": \\"key_hash\\"
                                }
                            ]
                        },
                        {
                            \\"prim\\": \\"Some\\",
                            \\"args\\": [
                                {
                                    \\"string\\": \\"tz1gTnKMA65qaKVpp6x4cgMLU2UyKF2zjHYN\\"
                                }
                            ]
                        }
                    ]
                },
                {
                    \\"prim\\": \\"SET_DELEGATE\\"
                },
                {
                    \\"prim\\": \\"CONS\\"
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Delegation operation Build operation to clear the delegation (Sugared version) 1`] = `
"
        (
            template_id (static_id 0 (\\"operation.test.ts\\" 95))
            storage (unit)
            storage_type (())
            messages ((ep1 True False False False (\\"operation.test.ts\\" 98) ((set (operations (\\"operation.test.ts\\" 98)) (cons (set_delegate (variant \\"None\\" (unit) (\\"operation.test.ts\\" 98)) (\\"operation.test.ts\\" 98)) (operations (\\"operation.test.ts\\" 98)) (\\"operation.test.ts\\" 98)) (\\"operation.test.ts\\" 98)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"operation.test.ts\\" 95))
        )
        "
`;

exports[`Delegation operation Build operation to clear the delegation (Sugared version) 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\",
                \\"annots\\": [
                    \\"%ep1\\"
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"NONE\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"key_hash\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"SET_DELEGATE\\"
                },
                {
                    \\"prim\\": \\"CONS\\"
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Delegation operation Build operation to set delegator (Sugared version) 1`] = `
"
        (
            template_id (static_id 0 (\\"operation.test.ts\\" 82))
            storage (unit)
            storage_type (())
            messages ((ep1 True False False True (\\"operation.test.ts\\" 85) ((set_type (params (\\"operation.test.ts\\" 85)) \\"key_hash\\" (\\"operation.test.ts\\" 85)) (set (operations (\\"operation.test.ts\\" 85)) (cons (set_delegate (variant \\"Some\\" (params (\\"operation.test.ts\\" 85)) (\\"operation.test.ts\\" 85)) (\\"operation.test.ts\\" 85)) (operations (\\"operation.test.ts\\" 85)) (\\"operation.test.ts\\" 85)) (\\"operation.test.ts\\" 85)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"operation.test.ts\\" 82))
        )
        "
`;

exports[`Delegation operation Build operation to set delegator (Sugared version) 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"key_hash\\",
                \\"annots\\": [
                    \\"%ep1\\"
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"UNPAIR\\"
                },
                {
                    \\"prim\\": \\"SOME\\"
                },
                {
                    \\"prim\\": \\"SET_DELEGATE\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"SWAP\\"
                },
                {
                    \\"prim\\": \\"CONS\\"
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Transaction operation Build operation (Desugared version) 1`] = `
"
        (
            template_id (static_id 0 (\\"operation.test.ts\\" 9))
            storage (unit)
            storage_type (())
            messages ((ep1 True False False False (\\"operation.test.ts\\" 12) ((set (operations (\\"operation.test.ts\\" 12)) (cons (transfer (params (\\"operation.test.ts\\" 12)) (literal (mutez 100) (\\"operation.test.ts\\" 12)) (openVariant (contract default \\"unit\\" (literal (address tz1gTnKMA65qaKVpp6x4cgMLU2UyKF2zjHYN) (\\"operation.test.ts\\" 12)) (\\"operation.test.ts\\" 12)) \\"Some\\" (variant \\"Some\\" (literal (string \\"CONTRACT_NOT_FOUND\\") (\\"operation.test.ts\\" 12)) (\\"operation.test.ts\\" 12)) (\\"operation.test.ts\\" 12)) (\\"operation.test.ts\\" 12)) (operations (\\"operation.test.ts\\" 12)) (\\"operation.test.ts\\" 12)) (\\"operation.test.ts\\" 12)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"operation.test.ts\\" 9))
        )
        "
`;

exports[`Transaction operation Build operation (Desugared version) 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\",
                \\"annots\\": [
                    \\"%ep1\\"
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"UNPAIR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PUSH\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"address\\"
                        },
                        {
                            \\"string\\": \\"tz1gTnKMA65qaKVpp6x4cgMLU2UyKF2zjHYN\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"CONTRACT\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"unit\\"
                        }
                    ],
                    \\"annots\\": [
                        \\"%default\\"
                    ]
                },
                {
                    \\"prim\\": \\"IF_NONE\\",
                    \\"args\\": [
                        [
                            {
                                \\"prim\\": \\"PUSH\\",
                                \\"args\\": [
                                    {
                                        \\"prim\\": \\"option\\",
                                        \\"args\\": [
                                            {
                                                \\"prim\\": \\"string\\"
                                            }
                                        ]
                                    },
                                    {
                                        \\"prim\\": \\"Some\\",
                                        \\"args\\": [
                                            {
                                                \\"string\\": \\"CONTRACT_NOT_FOUND\\"
                                            }
                                        ]
                                    }
                                ]
                            },
                            {
                                \\"prim\\": \\"FAILWITH\\"
                            }
                        ],
                        []
                    ]
                },
                {
                    \\"prim\\": \\"PUSH\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"mutez\\"
                        },
                        {
                            \\"int\\": \\"100\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"DIG\\",
                    \\"args\\": [
                        {
                            \\"int\\": \\"3\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"TRANSFER_TOKENS\\"
                },
                {
                    \\"prim\\": \\"CONS\\"
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Transaction operation Build operation (Sugared version) 1`] = `
"
        (
            template_id (static_id 0 (\\"operation.test.ts\\" 28))
            storage (unit)
            storage_type (())
            messages ((ep1 True False False False (\\"operation.test.ts\\" 31) ((set (operations (\\"operation.test.ts\\" 31)) (cons (transfer (unit) (literal (mutez 100) (\\"operation.test.ts\\" 31)) (openVariant (contract default \\"unit\\" (literal (address tz1gTnKMA65qaKVpp6x4cgMLU2UyKF2zjHYN) (\\"operation.test.ts\\" 31)) (\\"operation.test.ts\\" 31)) \\"Some\\" (variant \\"Some\\" (literal (string \\"CONTRACT_NOT_FOUND\\") (\\"operation.test.ts\\" 31)) (\\"operation.test.ts\\" 31)) (\\"operation.test.ts\\" 31)) (\\"operation.test.ts\\" 31)) (operations (\\"operation.test.ts\\" 31)) (\\"operation.test.ts\\" 31)) (\\"operation.test.ts\\" 31)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"operation.test.ts\\" 28))
        )
        "
`;

exports[`Transaction operation Build operation (Sugared version) 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\",
                \\"annots\\": [
                    \\"%ep1\\"
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"CDR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PUSH\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"address\\"
                        },
                        {
                            \\"string\\": \\"tz1gTnKMA65qaKVpp6x4cgMLU2UyKF2zjHYN\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"CONTRACT\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"unit\\"
                        }
                    ],
                    \\"annots\\": [
                        \\"%default\\"
                    ]
                },
                {
                    \\"prim\\": \\"IF_NONE\\",
                    \\"args\\": [
                        [
                            {
                                \\"prim\\": \\"PUSH\\",
                                \\"args\\": [
                                    {
                                        \\"prim\\": \\"option\\",
                                        \\"args\\": [
                                            {
                                                \\"prim\\": \\"string\\"
                                            }
                                        ]
                                    },
                                    {
                                        \\"prim\\": \\"Some\\",
                                        \\"args\\": [
                                            {
                                                \\"string\\": \\"CONTRACT_NOT_FOUND\\"
                                            }
                                        ]
                                    }
                                ]
                            },
                            {
                                \\"prim\\": \\"FAILWITH\\"
                            }
                        ],
                        []
                    ]
                },
                {
                    \\"prim\\": \\"PUSH\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"mutez\\"
                        },
                        {
                            \\"int\\": \\"100\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"UNIT\\"
                },
                {
                    \\"prim\\": \\"TRANSFER_TOKENS\\"
                },
                {
                    \\"prim\\": \\"CONS\\"
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;

exports[`Transaction operation Build operation to an originated contract 1`] = `
"
        (
            template_id (static_id 0 (\\"operation.test.ts\\" 41))
            storage (unit)
            storage_type (())
            messages ((ep1 True False False True (\\"operation.test.ts\\" 44) ((set_type (params (\\"operation.test.ts\\" 44)) \\"nat\\" (\\"operation.test.ts\\" 46)) (set (operations (\\"operation.test.ts\\" 46)) (cons (transfer (params (\\"operation.test.ts\\" 44)) (literal (mutez 100) (\\"operation.test.ts\\" 46)) (openVariant (contract ep1 \\"nat\\" (literal (address KT1R9M3MDffw7qSVSnbJs46aMC9YzzZz3aGT) (\\"operation.test.ts\\" 46)) (\\"operation.test.ts\\" 46)) \\"Some\\" (variant \\"Some\\" (literal (string \\"CONTRACT_NOT_FOUND\\") (\\"operation.test.ts\\" 46)) (\\"operation.test.ts\\" 46)) (\\"operation.test.ts\\" 46)) (\\"operation.test.ts\\" 46)) (operations (\\"operation.test.ts\\" 46)) (\\"operation.test.ts\\" 46)) (\\"operation.test.ts\\" 46)))))
            flags ()
            privates ()
            views ()
            entry_points_layout ()
            initial_metadata ()
            balance (literal (mutez 0) (\\"operation.test.ts\\" 41))
        )
        "
`;

exports[`Transaction operation Build operation to an originated contract 2`] = `
"[
    {
        \\"prim\\": \\"storage\\",
        \\"args\\": [
            {
                \\"prim\\": \\"unit\\"
            }
        ]
    },
    {
        \\"prim\\": \\"parameter\\",
        \\"args\\": [
            {
                \\"prim\\": \\"nat\\",
                \\"annots\\": [
                    \\"%ep1\\"
                ]
            }
        ]
    },
    {
        \\"prim\\": \\"code\\",
        \\"args\\": [
            [
                {
                    \\"prim\\": \\"UNPAIR\\"
                },
                {
                    \\"prim\\": \\"NIL\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"operation\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"PUSH\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"address\\"
                        },
                        {
                            \\"string\\": \\"KT1R9M3MDffw7qSVSnbJs46aMC9YzzZz3aGT\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"CONTRACT\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"nat\\"
                        }
                    ],
                    \\"annots\\": [
                        \\"%ep1\\"
                    ]
                },
                {
                    \\"prim\\": \\"IF_NONE\\",
                    \\"args\\": [
                        [
                            {
                                \\"prim\\": \\"PUSH\\",
                                \\"args\\": [
                                    {
                                        \\"prim\\": \\"option\\",
                                        \\"args\\": [
                                            {
                                                \\"prim\\": \\"string\\"
                                            }
                                        ]
                                    },
                                    {
                                        \\"prim\\": \\"Some\\",
                                        \\"args\\": [
                                            {
                                                \\"string\\": \\"CONTRACT_NOT_FOUND\\"
                                            }
                                        ]
                                    }
                                ]
                            },
                            {
                                \\"prim\\": \\"FAILWITH\\"
                            }
                        ],
                        []
                    ]
                },
                {
                    \\"prim\\": \\"PUSH\\",
                    \\"args\\": [
                        {
                            \\"prim\\": \\"mutez\\"
                        },
                        {
                            \\"int\\": \\"100\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"DIG\\",
                    \\"args\\": [
                        {
                            \\"int\\": \\"3\\"
                        }
                    ]
                },
                {
                    \\"prim\\": \\"TRANSFER_TOKENS\\"
                },
                {
                    \\"prim\\": \\"CONS\\"
                },
                {
                    \\"prim\\": \\"PAIR\\"
                }
            ]
        ]
    }
]"
`;
